gobblin-compaction/src/main/java/gobblin/compaction/mapreduce/MRCompactor.java

```diff
@@ -0,0 +1,314 @@
+
+  private final State state;
+  private final String inputDir;
+  private final String inputSubDir;
+  private final String destDir;
+  private final String destSubDir;
+  private final String tmpDir;
+  private final FileSystem fs;
+  private final ExecutorService executorService;
+  private final List<Future<?>> futures;
+
+  public MRCompactor(Properties props) throws IOException {
+    this.state = new State();
+    state.addAll(props);
+    this.inputDir = getInputDir();
+    this.inputSubDir = getInputSubDir();
+    this.destDir = getDestDir();
+    this.destSubDir = getDestSubDir();
+    this.tmpDir = getTmpDir();
+    this.fs = getFileSystem();
+    this.executorService = createExecutorService();
+    this.futures = Lists.newArrayList();
+  }
+
+  private String getInputDir() {
+    Preconditions.checkArgument(this.state.contains(ConfigurationKeys.COMPACTION_INPUT_DIR), "Property "
+        + ConfigurationKeys.COMPACTION_INPUT_DIR + " not provided");
+    return this.state.getProp(ConfigurationKeys.COMPACTION_INPUT_DIR);
+  }
+
+  private String getInputSubDir() {
+    return this.state.getProp(ConfigurationKeys.COMPACTION_INPUT_SUBDIR,
+        ConfigurationKeys.DEFAULT_COMPACTION_INPUT_SUBDIR);
+  }
+
+  private String getDestDir() {
+    Preconditions.checkArgument(this.state.contains(ConfigurationKeys.COMPACTION_DEST_DIR), "Property "
+        + ConfigurationKeys.COMPACTION_DEST_DIR + " not provided");
+    return this.state.getProp(ConfigurationKeys.COMPACTION_DEST_DIR);
+  }
+
+  private String getDestSubDir() {
+    return this.state.getProp(ConfigurationKeys.COMPACTION_DEST_SUBDIR,
+        ConfigurationKeys.DEFAULT_COMPACTION_DEST_SUBDIR);
+  }
+
+  private String getTmpDir() {
+    return this.state.getProp(ConfigurationKeys.COMPACTION_TMP_DIR, ConfigurationKeys.DEFAULT_COMPACTION_TMP_DIR);
+  }
+
+  private FileSystem getFileSystem() throws IOException {
+    Configuration conf = new Configuration();
+    for (String propName : this.state.getPropertyNames()) {
+      conf.set(propName, this.state.getProp(propName));
+    }
+    return FileSystem.get(conf);
+  }
+
+  private ExecutorService createExecutorService() {
+    int threadPoolSize = getThreadPoolSize();
+    return new ThreadPoolExecutor(threadPoolSize, threadPoolSize, 0L, TimeUnit.SECONDS,
+        new LinkedBlockingQueue<Runnable>());
+  }
+
+  private int getThreadPoolSize() {
+    return this.state.getPropAsInt(ConfigurationKeys.COMPACTION_THREAD_POOL_SIZE,
+        ConfigurationKeys.DEFAULT_COMPACTION_THREAD_POOL_SIZE);
+  }
+
+  @Override
+  public void compact() throws IOException {
+    try {
+      Set<String> topics = findAllTopics();
+      processTopics(topics);
+
+    } finally {
+      this.executorService.shutdown();
+      try {
+        executorService.awaitTermination(Long.MAX_VALUE, TimeUnit.SECONDS);
+        for (Future<?> future : this.futures) {
+          future.get();
```

Comment:
you are getting the results of the futures but not doing anything with them. You could skip this line. (That way a real timeout could exist).